File: divide_array_into_arrays_with_max_difference.go
func divideArray(nums []int, k int) [][]int {
    var ans [][]int
    sort.Ints(nums)

    for i := 2; i < len(nums); i = i + 3 {
        if nums[i]-nums[i-2] > k {
            return nil
        }
        ans = append(ans, []int{nums[i-2], nums[i-1], nums[i]})
    }
    return ans
}

// Time Complexity: O(nlogn)
// Space Complexity: O(n)

File: divide_array_into_arrays_with_max_difference.cpp
class Solution {
public:
    vector<vector<int>> divideArray(vector<int>& nums, int k) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> ans;

        for(int i=2; i < nums.size(); i=i+3) {
            if(nums[i]- nums[i-2] > k) return {};
            ans.push_back({nums[i-2], nums[i-1], nums[i]});
        }
        return ans;
    }
};

// Time Complexity: O(nlogn)
// Space Complexity: O(1)
